text/0005-syncing.md

```diff
@@ -0,0 +1,114 @@
+
+Given all the above:
+
+> Active node needs to know whole DAGs of all active (bonded) eras to correctly participate in the protocol.
+
+## How to synchronize ##
+
+### Building up the initial (Global) state ###
+
+Before we start accepting consensus messages we need to learn what are the current era validators – this information lives in the Global State. We also need the Global State if we want to execute deploys. 
+
+A node is started with an initial hash of the linear chain block (hash of the finalized block) that it gets from a trusted source (a friend, blockchain explorer, CasperLabs organization, newspaper). This block needs to be younger than the unbonding period (era in which it was created is still active). Otherwise, node is not guaranteed to receive consensus messages that will allow it advance the chain. Also, we cannot trust messages created in already inactive (unbonded) eras.
+
+#### Prerequisites ####
+
+1. Add `era_id` to `BlockHeader`
+2. Add `switch_block` flag to `BlockHeader`
+3. (optional) Add next-era validator set to the `BlockHeader` if`block.switch_block = true`.
+
+Normal way of operation of an active node is that Consensus component will send `FinalizedBlock` messages to the `BlockExecutor`. `FinalizedBlock` message has fields like:
+
+* `switch_block` flag,
+* `random_bit`
+* `era_id` (doesn't yet have but it should)
+* `Vec<SystemTransaction>`
+
+#### Variant A ####
+
+Once we receive the finalized block we started the node with, we check if we have all of its dependencies (parent block, deploys) and if not we ask peers for them. This process is repeated recursively until we have downloaded all of the dependencies and we can start executing deploys from the linear chain (starting from the oldest one).
+
+**Necessary work:**
+
+1. Start a node with the trusted hash.
+2. Ask peers for the initial block from the linear chain.
+3. Synchronize all missing dependencies.
+4. Execute deploys from the linear chain in the order of them appearing in the linear chain.
+
+`BlockExecutor` sees whether it's a `switch_block` and whether it should run an auction for era `era_id + AUCTION_DELAY` . As a result of execution, it will produce a `Block` instance (that is a block of the linear chain) but also an event `NextEraValidators(finalized_block.era_id + 1, Set<Validator>)` that informs Consensus about the next era validators. Consensus component will use new validators (and random bits from blocks between booking block and key block of the next era) and create an instance of new era.
+
+Same mechanism can be used in a node that is catching up – it will use `Block`s from the linear chain to build up its Global State via `BlockExecutor` (we need a mechanism of turning `Block` into `FinalizedBlock` as that's the input that `BlockExecutor` understands) and create new eras. 
```

Comment:
I was thinking that maybe instead of using parent's `ProtoBlock` hash we could use block height. This should be the same on all nodes and can be used to get block's `pre_state_hash`.

Reply:
We didn't observe that all nodes had the same block height for a given block hash in the Scala node. What is different in Rust impl. that makes this so?

Reply:
It's the linear chain – a chain of finalized blocks – and they all have the same block height. They had in Scala too (since all nodes have to finalize the same thing in the same order).

Reply:
That's true; unless there is an equivocation catastrophe — which we'd detect locally before finalizing any two blocks at the same height — the height should suffice.